src/fsharp/utils/FileSystem.fs

```diff
@@ -786,52 +806,90 @@ type internal ByteBuffer =
         buf.bbCurrent <- newSize
 
     member bb.FixupInt32 pos value =
+        bb.CheckDisposed()
         bb.bbArray.[pos] <- (Bytes.b0 value |> byte)
         bb.bbArray.[pos + 1] <- (Bytes.b1 value |> byte)
         bb.bbArray.[pos + 2] <- (Bytes.b2 value |> byte)
         bb.bbArray.[pos + 3] <- (Bytes.b3 value |> byte)
 
     member buf.EmitInt32 n =
+        buf.CheckDisposed()
         let newSize = buf.bbCurrent + 4
         buf.Ensure newSize
         buf.FixupInt32 buf.bbCurrent n
         buf.bbCurrent <- newSize
 
     member buf.EmitBytes (i:byte[]) =
+        buf.CheckDisposed()
         let n = i.Length
         let newSize = buf.bbCurrent + n
         buf.Ensure newSize
         Bytes.blit i 0 buf.bbArray buf.bbCurrent n
         buf.bbCurrent <- newSize
 
+    member buf.EmitMemory (i:ReadOnlyMemory<byte>) =
+        buf.CheckDisposed()
+        let n = i.Length
+        let newSize = buf.bbCurrent + n
+        buf.Ensure newSize
+        i.CopyTo(Memory(buf.bbArray, buf.bbCurrent, n))
+        buf.bbCurrent <- newSize
+
     member buf.EmitByteMemory (i:ReadOnlyByteMemory) =
+        buf.CheckDisposed()
         let n = i.Length
         let newSize = buf.bbCurrent + n
         buf.Ensure newSize
         i.Copy(0, buf.bbArray, buf.bbCurrent, n)
         buf.bbCurrent <- newSize
 
+    member buf.EmitByteBuffer (i:ByteBuffer) =
+        buf.CheckDisposed()
+        let n = i.Position
+        let newSize = buf.bbCurrent + n
+        buf.Ensure newSize
+        Bytes.blit i.bbArray 0 buf.bbArray buf.bbCurrent n
+        buf.bbCurrent <- newSize
+
     member buf.EmitInt32AsUInt16 n =
+        buf.CheckDisposed()
         let newSize = buf.bbCurrent + 2
         buf.Ensure newSize
         buf.bbArray.[buf.bbCurrent] <- (Bytes.b0 n |> byte)
         buf.bbArray.[buf.bbCurrent + 1] <- (Bytes.b1 n |> byte)
         buf.bbCurrent <- newSize
 
-    member buf.EmitBoolAsByte (b:bool) = buf.EmitIntAsByte (if b then 1 else 0)
+    member buf.EmitBoolAsByte (b:bool) = 
+        buf.CheckDisposed()
+        buf.EmitIntAsByte (if b then 1 else 0)
 
-    member buf.EmitUInt16 (x:uint16) = buf.EmitInt32AsUInt16 (int32 x)
+    member buf.EmitUInt16 (x:uint16) = 
+        buf.CheckDisposed()
+        buf.EmitInt32AsUInt16 (int32 x)
 
     member buf.EmitInt64 x =
+        buf.CheckDisposed()
         buf.EmitInt32 (Bytes.dWw0 x)
         buf.EmitInt32 (Bytes.dWw1 x)
 
-    member buf.Position = buf.bbCurrent
+    member buf.Position =
+        buf.CheckDisposed()
+        buf.bbCurrent
 
-    static member Create sz =
-        { bbArray = Bytes.zeroCreate sz
+    static member Create(sz, useArrayPool) =
+        { useArrayPool = defaultArg useArrayPool false
+          isDisposed = false
+          bbArray = ArrayPool.Shared.Rent sz
```

Comment:
This should depend on `useArrayPool` if I'm reading the intent right?

Reply:
Yes, that is right